kore/src/Kore/Attribute/Pattern/ConstructorLike.hs

```diff
@@ -0,0 +1,230 @@
+{- |
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+
+ -}
+
+module Kore.Attribute.Pattern.ConstructorLike
+    ( ConstructorLike (..)
+    , ConstructorLikeHead (..)
+    ) where
+
+import Control.DeepSeq
+import Data.Hashable
+import Data.Maybe
+    ( isJust
+    )
+import qualified Generics.SOP as SOP
+import qualified GHC.Generics as GHC
+
+import Kore.Attribute.Synthetic
+import Kore.Debug
+import Kore.Domain.Builtin
+import Kore.Internal.Alias
+    ( Alias
+    )
+import Kore.Internal.InternalBytes
+    ( InternalBytes
+    )
+import Kore.Internal.Symbol
+    ( Symbol
+    )
+import qualified Kore.Internal.Symbol as Symbol
+import Kore.Syntax
+import Kore.Syntax.Application
+    ( Application (..)
+    )
+import Kore.Variables.UnifiedVariable
+    ( UnifiedVariable
+    )
+
+{- | A pattern is 'ConstructorLike' if logical equality is syntactic equality.
+
+In other words, a pattern is constructor-like if it is equal (in the logical
+'Equals' sense) to another constructor-like pattern if and only if it is
+syntactically equal (in the 'Eq' sense).
+
+Examples of patterns that are constructor-like:
+- 'BuiltinBool', 'BuiltinInt', 'BuiltinString', and 'InternalBytes'
```

Comment:
Should a BuiltinMap with only constructor-like things in it be constructor-like?